base/applications/cmdutils/fc/fc.c

```diff
@@ -0,0 +1,498 @@
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Comparing files
+ * COPYRIGHT:   Copyright 2021 Katayama Hirofumi MZ (katayama.hirofumi.mz@gmail.com)
+ */
+#include <stdlib.h>
+#include <string.h>
+#include <windef.h>
+#include <winbase.h>
+#include <winnls.h>
+#include <conutils.h>
+#include "resource.h"
+
+// See also: https://stackoverflow.com/questions/33125766/compare-files-with-a-cmd
+typedef enum FCRET // return code of FC command
+{
+    FCRET_INVALID = -1, FCRET_IDENTICAL, FCRET_DIFFERENT, FCRET_CANT_FIND
+} FCRET;
+
+#define MAX_VIEW_SIZE (64 * 1024 * 1024) // 64 MB
+
+static __inline DWORD LOLONG(DWORDLONG dwl)
+{
+    ULARGE_INTEGER uli;
+    uli.QuadPart = dwl;
+    return uli.LowPart;
+}
+static __inline DWORD HILONG(DWORDLONG dwl)
+{
+    ULARGE_INTEGER uli;
+    uli.QuadPart = dwl;
+    return uli.HighPart;
+}
+
+#define FLAG_A (1 << 0)
+#define FLAG_B (1 << 1)
+#define FLAG_C (1 << 2)
+#define FLAG_L (1 << 3)
+#define FLAG_LBn (1 << 4)
+#define FLAG_N (1 << 5)
+#define FLAG_OFFLINE (1 << 6)
+#define FLAG_T (1 << 7)
+#define FLAG_U (1 << 8)
+#define FLAG_W (1 << 9)
+#define FLAG_nnnn (1 << 10)
+#define FLAG_HELP (1 << 11)
+
+typedef struct FILECOMPARE
+{
+    DWORD dwFlags; // FLAG_...
+    INT n, nnnn;
+    LPCWSTR file1, file2;
+} FILECOMPARE;
+
+static FCRET NoDifference(VOID)
+{
+    ConResPuts(StdOut, IDS_NO_DIFFERENCE);
+    return FCRET_IDENTICAL;
+}
+
+static FCRET Different(LPCWSTR file1, LPCWSTR file2)
+{
+    ConResPrintf(StdOut, IDS_DIFFERENT, file1, file2);
+    return FCRET_DIFFERENT;
+}
+
+static FCRET LongerThan(LPCWSTR file1, LPCWSTR file2)
+{
+    ConResPrintf(StdOut, IDS_LONGER_THAN, file1, file2);
+    return FCRET_DIFFERENT;
+}
+
+static FCRET OutOfMemory(VOID)
+{
+    ConResPuts(StdErr, IDS_OUT_OF_MEMORY);
+    return FCRET_INVALID;
+}
+
+static FCRET CannotRead(LPCWSTR file)
+{
+    ConResPrintf(StdErr, IDS_CANNOT_READ, file);
+    return FCRET_INVALID;
+}
+
+static FCRET InvalidSwitch(VOID)
+{
+    ConResPuts(StdErr, IDS_INVALID_SWITCH);
+    return FCRET_INVALID;
+}
+
+static VOID CannotOpen(LPCWSTR file, DWORD dwError)
+{
+    LPVOID lpMsgBuf;
+    DWORD dwFlags = FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM |
+                    FORMAT_MESSAGE_IGNORE_INSERTS;
+    FormatMessageW(dwFlags, NULL, dwError, 0, (LPWSTR)&lpMsgBuf, 0, NULL);
+    ConResPrintf(StdErr, IDS_CANNOT_OPEN, file, (LPWSTR)lpMsgBuf);
+    LocalFree(lpMsgBuf);
+}
+
+static HANDLE DoOpenFileForInput(LPCWSTR file)
+{
+    HANDLE hFile = CreateFileW(file, GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, 0, NULL);
+    if (hFile == INVALID_HANDLE_VALUE)
+        CannotOpen(file, GetLastError());
+    return hFile;
+}
+
+static BOOL GetFileSizeDx(HANDLE hFile, size_t *pcbFile, BOOL bLargeOK)
+{
+    LARGE_INTEGER li;
+    if (!GetFileSizeEx(hFile, &li) || (!bLargeOK && li.QuadPart >= MAXLONG))
+        return FALSE;
+    *pcbFile = (size_t)li.QuadPart;
+    return TRUE;
+}
```

Comment:
Please just use `GetFileSizeEx` with `LARGE_INTEGER`s instead of doing this game (there and in the other functions below where you use LOSIZE and HISIZE )

Reply:
OK.